kernel/src/engine/default/json.rs

```diff
@@ -303,4 +450,62 @@ mod tests {
         assert_eq!(data.len(), 1);
         assert_eq!(data[0].num_rows(), 4);
     }
+
+    #[tokio::test]
+    async fn test_read_json_files_ordering() {
+        let paths = [
+            "./tests/data/table-with-dv-small/_delta_log/00000000000000000000.json",
+            "./tests/data/table-with-dv-small/_delta_log/00000000000000000001.json",
+        ];
+        let paths = paths.map(|p| std::fs::canonicalize(PathBuf::from(p)).unwrap());
+
+        let path_string = paths[0].to_string_lossy().to_string();
+        let object_store_path = Path::from(path_string);
+        // for the first 000000.json, wait for 1 second
+        let key_map = (object_store_path, Duration::from_secs(1));
+
+        let store = Arc::new(SlowGetStore::new(
+            LocalFileSystem::new(),
+            vec![key_map].into_iter().collect(),
+        ));
+
+        let file_futures: Vec<_> = paths
+            .iter()
+            .map(|path| {
+                let store = store.clone();
+                async move {
+                    let url = url::Url::from_file_path(path).unwrap();
+                    let location = Path::from(url.path());
+                    let meta = store.head(&location).await.unwrap();
+                    FileMeta {
+                        location: url.clone(),
+                        last_modified: meta.last_modified.timestamp_millis(),
+                        size: meta.size,
+                    }
+                }
+            })
+            .collect();
+        let files = future::join_all(file_futures).await;
+
+        let handler = DefaultJsonHandler::new(store, Arc::new(TokioBackgroundExecutor::new()));
```

Comment:
`TokioBackgroundExecutor` does: `let rt = tokio::runtime::Builder::new_current_thread()`. tokio's `sleep` [is not guaranteed to yield](https://users.rust-lang.org/t/does-awaiting-tokio-sleep-duration-zero-yield-back-to-the-runtime-or-poll-ready-immediately/116756). 

Probably we should use `TokioMultiThreadExecutor` to ensure we get things out of order at least sometimes.

Reply:
ah good catch - I was thinking it's safe to assume a 1 second sleep would yield but just to be safe will change :)